runtime/compiler/optimizer/J9RecognizedCallTransformer.cpp

```diff
@@ -692,6 +694,33 @@ void J9::RecognizedCallTransformer::processUnsafeAtomicCall(TR::TreeTop* treetop
       if (enableTrace)
          traceMsg(comp(), "Created isObjectNull test node n%dn, non-null object will fall through to Block_%d\n", isObjectNullNode->getGlobalIndex(), treetop->getEnclosingBlock()->getNumber());
 
+      // Test if object is array - offheap only
+   #if defined (J9VM_GC_ENABLE_SPARSE_HEAP_ALLOCATION)
+      if (TR::Compiler->om.isOffHeapAllocationEnabled() && comp()->target().is64Bit())
+         {
+         //generate array check treetop
+         TR::Node *constNode = TR::Node::lconst(unsafeCall, ~(J9_SUN_FIELD_OFFSET_MASK));
+         TR::Node *unsafeAddress = TR::Node::create(TR::aladd, 2, objectNode, TR::Node::create(TR::land, 2, offsetNode, constNode));
+         TR::Node *vftLoad = TR::Node::createWithSymRef(TR::aloadi, 1, 1,
+                                                        TR::Node::createWithSymRef(unsafeAddress,
+                                                                                   comp()->il.opCodeForDirectLoad(unsafeAddress->getDataType()),
+                                                                                   0,
+                                                                                   objectNode->getSymbolReference()),
+                                                        comp()->getSymRefTab()->findOrCreateVftSymbolRef());
```

Comment:
don't understand this either:  you seemed assuming the previous directload result as an object.  at the same time, you look up the opcode through a table (while you know the datatype already).